### ISSUE TITLE
pghistory adds NEW data not OLD, requiring a total database copy for robust backups

### ISSUE THREAD

**Issue by mlissner** [Opus10/django-pghistory#78]
Until I looked under the hood, my mental model for pghistory was that it saved the OLD value whenever a row got an `UPDATE` or **`DELETE`**. That way, you could always revert. 

It turns out that it saves a copy of the NEW value instead, whenever the row gets an `UPDATE` or **`INSERT`**.

This seems to work OK, but it seems to leave a big pathway for major problems:

1. Create a model, add data to it. (`INSERT`)

2. Add pghistory to track the model.

3. Change the data. (`UPDATE` new data in the model; `INSERT` copy of new data into the event table)

4. Realize the change was bad; try to revert.

5. 💥 You can't revert. The original is gone. The event table holds the NEW data, not the OLD.  💥

Maybe I'm missing something, but if you don't have pghistory enabled from the beginning of a project, you can't revert to versions of data that existed when you started using pghistory. Is that right?

I think there are two solutions to this:

1. Save the OLD value to the event table, not the new, whenever something is UPDATED or DELETED.

2. Run some sort of script to save copies of the current data whenever you add pghistory to a model. This isn't great if you have a very big database like we do, since it means you have to copy the whole thing on the off chance a row is changed.

Either way, I looked through the docs a bit and I didn't see any warnings about this. It feels like it should be a big, loud warning in a couple places (FAQ, Homepage, etc.), or maybe an option when setting up triggers.

Is there a reason for the current design? Sorry if I missed something in the docs and thank you again for the great project!

**Comment by mlissner** [comment]
Hi, I'm just coming around to this because I think we're going to tweak the way we use pghistory from the default decorators, and I thought I'd check one more time if any of the creators of the project knew why it saves `NEW` instead of `OLD`. 

If there's no good reason, I'd like to file a documentation update at least, to highlight this problem and offer a workaround, once we have it figured out. I'm not a DBA and I'm not deep on this topic though, so I'm pretty hesitant to do that until somebody with some expertise chime in. I feel like I must be missing the advantage of the current design.

**Comment by mlissner** [comment]
This issue hasn't gotten any feedback, so maybe I'm wasting my time, but we're moving ahead (I think!) on a tracker that works for UPDATE or DELETE and that inserts the OLD data into the event model. 

Would this be welcome as a new feature? If so, we'd be very pleased to do a PR with it so that others can benefit from this approach as well. It's based heavily off of the base `Snapshot` class. 

You can see a PR of that in our code here (though it needs a few more tweaks): https://github.com/freelawproject/courtlistener/pull/2593/files#diff-a1a6e0c0974b7e350ee7debb9f20bbdd69162d6e2be7ed8eb2961e3fe0b75a91

Thanks!

**Comment by wesleykendall** [comment]
This one is tough to answer.

It seems like the question here is if you should snapshot NEW on INSERT, right? If you perform your strategy of snapshotting OLD on UPDATE, you would not want to snapshot NEW on INSERT, otherwise you'd have duplicate snapshots.

With that, let's assume we don't snapshot on INSERT. Here are some of the problems that would happen:

1. You can't easily render a diff of the most recent changes to rows. You would have to diff the tracked model and the most recent snapshot model, which would be really tricky to support in this library since the snapshot model can have a totally different schema.
2. You would lose context information about who created the model. For example, if you use the pghistory middleware and attach context to changes, you would not know which user created the initial snapshot of the model.

These are the main reasons I used this as the default behavior. You do bring up good points about needing to backfill data though. And there's also a problem of not snapshotting on deletion since you would also lose contextual data here. It's a tough problem because of how pghistory models it, and there are tradeoffs in both scenarios.

Hope that explanation makes sense, and I hope you understand what impacts you'll have in your application if you use context tracking or the admin. 

It seems like the way forward are a few possibilities:

1. I agree there should be something in the docs that lets users know that existing rows will have no history attached to them until they're updated. 
2. I think there's room for a utility to populate history of existing rows. It would be nice if it was automatically added in migrations, but at least it could be a migration operation users could add manually if they want.
3. There might be room for another core tracker that snapshots OLD on UPDATE and DELETE so that you don't have to add multiple trackers for this use case. I would still keep this as a more advanced use case though with warnings in the docs since it fundamentally alters diffing and context tracking. (i.e. users will be confused when the context around who created a model isn't tracked).
4. Finally, there might be room for a solution where you do snapshot NEW on INSERT, OLD on UPDATE, and OLD on DELETE (or some other mix). I just really don't know what that is at the moment without a lot of confusion and complexity.

I hope this explanation helps clear some things up.

TL:DR; If you don't snapshot NEW on INSERT, you'll lose context associated with who created a row. You'll also lose the ability to easily show the most recent diffs of changes. I think there should be a warning in the docs about existing rows and possibly a utility to migrate existing rows.

Let me know what you think!

**Comment by mlissner** [comment]
> Hope that explanation makes sense,

Yes, those tradeoffs are really clear now, thank you. For our part, we don't use the middlware or context, generate diffs, or plan to use the history except as a backup/audit table, so these tradeoffs are totally fine for us.

> I agree there should be something in the docs that lets users know that existing rows will have no history attached to them until they're updated.

We'll definitely do this. I think it fits best in the "Event Tracking" page.

> I think there's room for a utility to populate history of existing rows. It would be nice if it was automatically added in migrations, but at least it could be a migration operation users could add manually if they want.

We have close to a terabyte of data, so this isn't a solution we want to pursue, sorry.  I'm happy to create an issue for this though, if somebody wants to pick it up, and can link to it from the docs, if that's helpful.

> There might be room for another core tracker that snapshots OLD on UPDATE and DELETE so that you don't have to add multiple trackers for this use case.

Fantastic. We'll do a PR for this and hopefully it'll be good enough to merge. 

> Finally, there might be room for a solution where you .....

Yeah, I pondered this too, but I'm not seeing it either, at least not intuitively. Maybe? Seems utopian. Iit could give us everything and more if we can just figure it out....

> I hope this explanation helps clear some things up.

Yeah it does! Thank you! We'll get to work!

**Comment by wesleykendall** [comment]
Check #83 to see the temporary resolution for this (specifically [this comment](https://github.com/Opus10/django-pghistory/issues/83#issuecomment-1500932552) ), which is by manually creating a new snapshot class. As mentioned, using this strategy of UPDATE/DELETE does cause some possible side effects with the admin and other things.

I think long term I'm going to try to consolidate the trackers a bit better, but it would require breaking the API. I'm going to keep this issue open until I update the docs

**Comment by wesleykendall** [comment]
To follow up once more, the trackers have been consolidated better in [version 3](https://django-pghistory.readthedocs.io/en/3.3.0/release_notes/#300-2023-10-16). I'm hopeful that this combined with [some of the new conditions](https://django-pgtrigger.readthedocs.io/en/4.12.0/conditional/) and ability to set default configuration with `PGHISTORY_DEFAULT_TRACKERS` can help with removing boilerplate for different default setups.

Feel free to open a [discussion](https://github.com/Opus10/django-pghistory/discussions) if you experience any awkwardness with the new interface or have suggestions!

**Comment by mlissner** [comment]
Amazing, we'll check this out!
